nautobot/extras/datasources/git.py

```diff
@@ -204,6 +287,46 @@ def ensure_git_repository(repository_record, job_result=None, logger=None, head=
         logger.info("Repository successfully refreshed")
 
 
+def git_repository_dry_run(repository_record, job_result=None, logger=None):
+    """Log the difference between local branch and remote branch files.
+
+    Args:
+        repository_record (GitRepository)
+        job_result (JobResult): Optional JobResult to store results into.
+        logger (logging.Logger): Optional Logger to additionally log results to.
+    """
+    from_url, to_path, from_branch = get_repo_from_url_to_path_and_from_branch(repository_record)
+
+    try:
+        repo_helper = GitRepo(to_path, from_url)
+        logger.info("Fetching from origin")
+        modified_files = repo_helper.diff_remote()
+
+        if modified_files:
+            # Log each modified files
+            for item in modified_files:
+                job_result.log(item, level_choice=LogLevelChoices.LOG_INFO, logger=logger)
+                job_result.save()
+        else:
+            job_result.log("Repository has no changes", level_choice=LogLevelChoices.LOG_INFO, logger=logger)
+            job_result.save()
+
+    except Exception as exc:
+        if job_result:
+            job_result.set_status(JobResultStatusChoices.STATUS_ERRORED)
+            job_result.log(str(exc), level_choice=LogLevelChoices.LOG_FAILURE, logger=logger)
+            job_result.save()
+        elif logger:
+            logger.error(str(exc))
+        raise
+
+    if job_result:
+        job_result.log("Repository Dry Run successful", level_choice=LogLevelChoices.LOG_SUCCESS, logger=logger)
+        job_result.save()
```

Comment:
No need any more to call `job_result.save()` after each `job_result.log()`, as `log()` now creates `JobLogEntry` records and doesn't actually modify the `JobResult`.